Wrapping unsafe raw JSContext pointers in a safe struct. Part 1.

marmeladema opened the pull request:
Wrapping unsafe raw JSContext pointers in a safe struct. Issue #20377.


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] create a struct JSContext(*mut jsapi::JSContext) type
- [x] implement Deref for this new type so it's easy to obtain the inner context pointer
- [x] add an unsafe from_ptr static method that returns a new instance of the type
- [x] start by changing various Argument uses in CodegenRust.py for python classes that inherit from CGAbstractMethod
- [x] convert the python code that interacts with CGClass (CGCallback, CGCallbackFunction, CGCallbackFunctionImpl, CGCallbackInterface) and any rust code that interacts with it
- [x] update CGDictionary and any rust code that interacts with it
- [x] make the code generator declare trait methods that accept the new type instead of *mut JSContext (CGInterfaceTrait)
- [x] modify GlobalScope::get_cx to return a SafeJSContext 

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/23816)
<!-- Reviewable:end -->

marmeladema committed:
Rename JSContext from js::jsapi to RawJSContext

marmeladema committed:
Add safe JSontext wrapper around unsafe RawJSContext

marmeladema committed:
ConstructorEnabled now takes a SafeJSContext instead of a JSContext

a first argument. The function cannot be made safe because of call
to unsafe function is_exposed_in.

marmeladema committed:
Wrap(Global)Method now takes a SafeJSContext instead of a JSContext

as first argument.

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @emilio (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/bindings/codegen/CodegenRust.py, components/script/dom/node.rs, components/script/dom/testworkletglobalscope.rs, components/script/dom/dedicatedworkerglobalscope.rs, components/script/dom/bindings/reflector.rs and 6 more
 * @kichjang: components/script/dom/bindings/codegen/CodegenRust.py, components/script/dom/node.rs, components/script/dom/testworkletglobalscope.rs, components/script/dom/dedicatedworkerglobalscope.rs, components/script/dom/bindings/reflector.rs and 6 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify script code, but no tests are modified. Please consider adding a test!

marmeladema committed:
CreateInterfaceObjects now takes a SafeJSContext instead of a JSContext

as first argument.

marmeladema committed:
GetPerInterfaceObject methods now takes a SafeJSContext instead of a JSContext

as first argument.

marmeladema committed:
DefineDOMInterfaceMethod now takes a SafeJSContext instead of a JSContext

as first argument.

jdm commented on components/script/dom/bindings/codegen/CodegenRust.py:
```diff
@@ -3131,7 +3131,7 @@ class CGGetPerInterfaceObject(CGAbstractMethod):
     """
     def __init__(self, descriptor, name, idPrefix="", pub=False):
-        args = [Argument('*mut JSContext', 'cx'),
+        args = [Argument('&mut SafeJSContext', 'cx'),
```
Why do we need a reference at all instead of `Argument('SafeJSContext')`?

marmeladema replied:
To not transfer ownership to the function? Its either that or deriving `Copy` for SafeJSContext i believe.

jdm replied:
Deriving Copy is something we want.

marmeladema replied:
Ok i'll update the code :+1:

jdm commented on components/script/script_runtime.rs:
```diff
@@ -748,3 +748,28 @@ unsafe fn set_gc_zeal_options(cx: *mut RawJSContext) {
+
+#[allow(unsafe_code)]
+impl Deref for JSContext {
+    type Target = RawJSContext;
```
This should be `*mut RawJSContext` instead. There should be no need to dereference the actual JSContext pointer as part of this operation.

jdm commented on components/script/script_runtime.rs:
```diff
@@ -748,3 +748,28 @@ unsafe fn set_gc_zeal_options(cx: *mut RawJSContext) {
+}
+
+#[allow(unsafe_code)]
+impl DerefMut for JSContext {
```
We should not need to implement DerefMut for this type.

marmeladema commented:
@jdm i updated the code according to your comments. I am heading toward the right direction here?

jdm commented:
Yes! This looks like what I expected to see :D

marmeladema commented:
I am having trouble modifying `CGAbstractStaticBindingMethod` because i end up having a bunch of errors like
```
error[E0308]: mismatched types
   --> /home/adema/code/servo/target/debug/build/script-537c159507a8c0bf/out/Bindings/ConsoleBinding.rs:808:42
    |
808 |         call: JSNativeWrapper { op: Some(time), info: 0 as *const JSJitInfo },
    |                                          ^^^^ expected *-ptr, found struct `script_runtime::JSContext`
    |
    = note: expected type `unsafe extern "C" fn(*mut js::jsapi::JSContext, u32, *mut js::jsapi::Value) -> bool`
               found type `unsafe extern "C" fn(script_runtime::JSContext, u32, *mut js::jsapi::Value) -> bool {dom::bindings::codegen::Bindings::ConsoleBinding::ConsoleBinding::time}`
```
which i don't know how to solve.

jdm commented:
Ah, indeed. Those will need to continue to have the raw pointer argument because the C function pointers can't be changed. We will want to declare a `let cx = SafeJSContext::from_ptr(cx)` as the first line of code in them instead.

marmeladema committed:
Callbacks now uses safe JSContext instead of raw JSContext

marmeladema committed:
Convert internal methods to handle safe JSContext instead of raw JSContext

jdm commented:
Looking good so far!

marmeladema committed:
Convert CGTraitInterface to use safe JSContext instead of raw JSContext

marmeladema commented:
@jdm i've followed the steps you described in the other PR https://github.com/servo/servo/pull/23657#issuecomment-510225906
We do you think are the next steps here?

jdm commented:
The failure does comefromyourchanges. If you run`/Mach test-tidy` you should reproduce it.

jdm commented:
I think the next big step would be modifying GlobalScope::get_cx to return a SafeJSContext and store one internally as well. That should remove a bunch of uses of from_ptr uses right now. At that point it would probably make sense to merge the changes and file follow-ups for increasing the usage in hand-written Rust code.

marmeladema commented:
> The failure does comefromyourchanges. If you run/Mach test-tidy you should reproduce it.

Ok it should be fixed now.

> I think the next big step would be modifying GlobalScope::get_cx to return a SafeJSContext and store one internally as well. That should remove a bunch of uses of from_ptr uses right now. At that point it would probably make sense to merge the changes and file follow-ups for increasing the usage in hand-written Rust code.

I'll look into it then :+1:

marmeladema committed:
Modify *::get_cx methods to return a safe JSContext instead of a raw one

marmeladema commented:
@jdm I made all methods `*::get_cx` return a safe `JSContext` but I did not store one inside `GlobalScope` since it does not store one itself and uses `Runtime::get()` instead.

jdm reviewed (changes requested):
Great work! Let's fix these two nits and merge it!

jdm commented:
I'm really happy with how this turned out. There are a bunch of pieces of code where the intention of the unsafety is much more clear now that we're not able to pass an arbitrary pointer as an argument. Thank you for finishing this first step, and thank you @oneturkmen for all the initial work that helped us figure out the best way to go about it!

oneturkmen commented:
@marmeladema Great job! :fire:, and @jdm thanks for the patience! :) 

Thanks to both of you for completing this - now I can see what I missed here and there.

Well done!

marmeladema commented on components/script/dom/bindings/htmlconstructor.rs:
```diff
@@ -109,31 +110,31 @@ where
             // Retrieve the constructor object for HTMLElement
             HTMLElementBinding::GetConstructorObject(
-                window.get_cx(),
+                SafeJSContext::from_ptr(*window.get_cx()),
```
Should be fixed now!

marmeladema commented on components/script/dom/bindings/reflector.rs:
```diff
@@ -16,14 +17,20 @@ use std::default::Default;
-    unsafe { wrap_fn(global_scope.get_cx(), global_scope, obj) }
+    unsafe {
+        wrap_fn(
+            JSContext::from_ptr(*global_scope.get_cx()),
```
Fixed too.

marmeladema commented:
Tahnk you @jdm for the good advices and @oneturkmen for paving the road :+1: 
I have to admit that I did not anticipated such a big amount of changes but it was fine at the end!

I am ready for the next part of refactoring if need be. I was thinking about a few things, one of them being maybe introducing intermediary `FromJSValConvertible` and `ToJSValConvertible` traits that would take a safe JSContext and automatically provide an implementation for the raw traits from `js::conversions`

jdm commented:
The intermediary traits aren't a bad idea! That and the remaining hand-written functions/methods that still accept `*mut JSContext` are some of the biggest causes of `*cx` bits in the code.

@bors-servo r+

bors-servo commented:
:pushpin: Commit 0de646f has been approved by `jdm`

<!-- @bors-servo r=jdm 0de646f482ae3d4bc5a62001890d07b6a58d61ff -->

bors-servo commented:
:lock: Merge conflict

marmeladema commented:
@jdm i rebased and fixed the conflicts but appveyor failed and i don't really understand what is the problem. I don't know what to look for in the build log.

jdm commented:
It hit the time limit for the build and it's very bad at making that clear. We can ignore it.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 88cacfb has been approved by `jdm`

<!-- @bors-servo r=jdm 88cacfb0098e20be70c27bfde6b74cd3290f1fe4 -->

bors-servo commented:
:hourglass: Testing commit 88cacfb0098e20be70c27bfde6b74cd3290f1fe4 with merge 9b14bbdc98ea078a4f868abee1b7c7244161f46d...

bors-servo commented:
:sunny: Test successful - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/12914), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/12974), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/ABbt43QfTiurCVvCnmprVg)
Approved by: jdm
Pushing 9b14bbdc98ea078a4f868abee1b7c7244161f46d to master...
